### ISSUE TITLE
Failing to silently login - auth fails on request - initialisation order problem?

### ISSUE THREAD

**Issue by wivlaro** [playgameservices/play-games-plugin-for-unity#2361]
I am having trouble with the silent login

I'm not really sure because I can't see the source of the GPGS Android SDK, but it looks like it's trying to authenticate before the SDK is initialised. We get this message in logcat, then an error "TokenFragment: No connected Games API", then 100ms later "GamesUnitySDK: Performing Android initialization of the GPG SDK"

The first attempt causes the silent auth callbacks to call and then I never get a result. The hacky fix I'm tempted to do is just try and login twice, or wait a few frames between initialisation and silent login, but that seems really wrong.

I want an id Token because I want to be able to verify that it is that user logged in on our servers. I'm using no other services apart from the authentication.

What's actually going on?

My startup code looks like this: 

```
            GooglePlayGames.OurUtils.Logger.DebugLogEnabled = true;
            var config = new PlayGamesClientConfiguration.Builder()
                .RequestIdToken()
                .Build();
            PlayGamesPlatform.InitializeInstance(config);
            PlayGamesPlatform.Activate();
            if (PlayGamesPlatform.Instance.IsAuthenticated())
            {
                Logger.InfoCh(ApeSocial.LOGCH, "Google Play Games already authenticated", PlayGamesPlatform.Instance.GetUserId());
            }
            else
            {
                var awaitingAuthentication = true;
                Logger.InfoCh(ApeSocial.LOGCH, "Starting Google Play Games auth");
                PlayGamesPlatform.Instance.Authenticate((success, message) =>
                {
                    Logger.InfoCh(ApeSocial.LOGCH, "Google Play Games auth result=",success?"succeeded":"failed","message:",message);
                    awaitingAuthentication = false;
                    if (success && handler != null)
                    {
                        handler.HandleLoggedIn(PlayGamesPlatform.Instance.localUser);
                    }
                }, true);
                var timeoutTime = DateTime.UtcNow.AddSeconds(loginTimeoutSeconds);
                while (awaitingAuthentication)
                {
                    if (DateTime.UtcNow >= timeoutTime)
                    {
                        Logger.WarnCh(ApeSocial.LOGCH, "Social login timeout");
                        break;
                    }
                    yield return null;
                }
            }
```

Here's an excerpt from my logcat
```
11-15 11:37:39.156 23330 23345 I Unity   : Starting Auth with token client.
11-15 11:37:39.156 23330 23345 I Unity   :  
11-15 11:37:39.156 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.156 23330 23345 I Unity   : 
11-15 11:37:39.186 23330 23345 W Unity   : !!! [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 WARNING: Creating new PlayGamesPlatform
11-15 11:37:39.186 23330 23345 W Unity   :  
11-15 11:37:39.186 23330 23345 W Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.186 23330 23345 W Unity   : 
11-15 11:37:39.187 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: Activating PlayGamesPlatform.
11-15 11:37:39.187 23330 23345 I Unity   :  
11-15 11:37:39.187 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.187 23330 23345 I Unity   : 
11-15 11:37:39.189 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: PlayGamesPlatform activated: GooglePlayGames.PlayGamesPlatform
11-15 11:37:39.189 23330 23345 I Unity   :  
11-15 11:37:39.189 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.189 23330 23345 I Unity   : 
11-15 11:37:39.189 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: Creating platform-specific Play Games client.
11-15 11:37:39.189 23330 23345 I Unity   :  
11-15 11:37:39.189 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.189 23330 23345 I Unity   : 
11-15 11:37:39.190 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: Creating Android IPlayGamesClient Client
11-15 11:37:39.190 23330 23345 I Unity   :  
11-15 11:37:39.190 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.190 23330 23345 I Unity   : 
11-15 11:37:39.213 23330 23345 D TokenFragment: Creating fragment
11-15 11:37:39.215 23330 23330 D TokenFragment: onStart()
11-15 11:37:39.215 23330 23330 D TokenFragment: onResume called
11-15 11:37:39.216 23330 23330 D TokenFragment: Building client for: a9eb6af (a:false e:false i:true wc: 913625813542-r7hjf3fm1prhf0k1v1ifqfa7hrcunru5.apps.googleusercontent.com f: false)
11-15 11:37:39.225 23330 23345 I Unity   : ---- [0] -- 97
11-15 11:37:39.225 23330 23345 I Unity   :  
11-15 11:37:39.225 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.225 23330 23345 I Unity   : 
11-15 11:37:39.250 23330 23330 W PopupManager: You have not specified a View to use as content view for popups. Falling back to the Activity content view. Note that this may not work as expected in multi-screen environments
11-15 11:37:39.285 23330 23330 D TokenFragment: No connected Games API
11-15 11:37:39.285 23330 23330 E TokenFragment: Setting result error code to: 13
11-15 11:37:39.356 23330 23330 D TokenPendingResult:  Calling onResult for callback: GooglePlayGames.Android.TokenResultCallback result: Status: Status{statusCode=ERROR, resolution=null} email: <null> id:<null> access: <null>
11-15 11:37:39.382 23330 23330 I Unity   : Fetch token result: 13: 
11-15 11:37:39.382 23330 23330 I Unity   :  
11-15 11:37:39.382 23330 23330 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.382 23330 23330 I Unity   : 
11-15 11:37:39.415 23330 23330 D GamesUnitySDK: Performing Android initialization of the GPG SDK
11-15 11:37:39.468 23330 23330 I Unity   : Building GPG services, implicitly attempts silent auth
11-15 11:37:39.468 23330 23330 I Unity   :  
11-15 11:37:39.468 23330 23330 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.468 23330 23330 I Unity   : 
11-15 11:37:39.471 23330 23330 I GamesNativeSDK: Using existing jar.
11-15 11:37:39.471 23330 23330 I GamesNativeSDK: Writing 1941 bytes to jar file
11-15 11:37:39.476 23330 23330 W zygote  : Skipping duplicate class check due to unrecognized classloader
11-15 11:37:39.511 23330 23330 I GamesNativeSDK: Using existing jar.
11-15 11:37:39.511 23330 23330 I GamesNativeSDK: Writing 1066 bytes to jar file
11-15 11:37:39.514 23330 23330 W zygote  : Skipping duplicate class check due to unrecognized classloader
11-15 11:37:39.528 23330 23330 W PopupManager: You have not specified a View to use as content view for popups. Falling back to the Activity content view. Note that this may not work as expected in multi-screen environments
11-15 11:37:39.529 23330 23476 I GamesNativeSDK: Auth operation started: SIGN IN
11-15 11:37:39.529 23330 23476 I GamesNativeSDK: Connecting to Google Play...
11-15 11:37:39.573 23330 23330 D TokenFragment: Done with processRequest, result is pending.
11-15 11:37:39.586 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: Starting Auth Transition. Op: SIGN_IN status: 13
11-15 11:37:39.586 23330 23345 I Unity   :  
11-15 11:37:39.586 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.586 23330 23345 I Unity   : 
11-15 11:37:39.587 23330 23345 I Unity   :  [Play Games Plugin DLL] 11/15/18 11:37:39 +00:00 DEBUG: Invoking callbacks, AuthState changed from silentPending to Unauthenticated.
11-15 11:37:39.587 23330 23345 I Unity   :  
11-15 11:37:39.587 23330 23345 I Unity   : (Filename: /Users/builduser/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)
11-15 11:37:39.587 23330 23345 I Unity   : 


```

**Comment by olehkuznetsov** [comment]
What version of the plugin do you use?

**Comment by wivlaro** [comment]
So as I'm digging and digging. Looking at TokenFragment, it seems that the TokenFragment is not appropriately waited for the API being connected.

Also it looks like the TokenFragment is passing through status codes as a result that have a different meaning; [this code](https://github.com/supersolid-Justy/play-games-plugin-for-unity/blob/c8e1383d50cf377a18dba4088689f14f9026e7f1/source/SupportLib/PlayGamesPluginSupport/src/main/java/com/google/games/bridge/TokenFragment.java#L529) passes through [these status codes](https://developers.google.com/android/reference/com/google/android/gms/auth/api/signin/GoogleSignInStatusCodes) into what ultimately is interpreted as AuthStatus [here](https://github.com/supersolid-Justy/play-games-plugin-for-unity/blob/c8e1383d50cf377a18dba4088689f14f9026e7f1/source/PluginDev/Assets/GooglePlayGames/Platforms/Native/NativeClient.cs#L137). Sometimes I was getting error codes 12501 which now make sense as I discover this.

It looks like the onResume is called at the start of the fragment, which calls processRequest, which doesn't wait for the GoogleApiClient to be connected before giving up and claiming it's a developer error? It passes through CommonStatusCodes.ERROR, which also ends up as AuthStatus and isn't covered there...

😭

**Comment by olehkuznetsov** [comment]
Yes, I see the possibility for onResume to be called before onConnected too.
Still not sure why I can't reproduce it locally.

**Comment by olehkuznetsov** [comment]
Should be fixed in v0.9.54
